peps/pep-9995.rst

```diff
@@ -0,0 +1,218 @@
+implemented as a function at runtime. To be made subscriptable, it will be
+changed to be a class.
+
+Creating an inlined typed dictionary results in a new class, so both syntaxes
+return the same type::
+
+    from typing import TypedDict
+
+    T1 = TypedDict('T1', {'a': int})
+    T2 = TypedDict[{'a': int}]
+
+
+Backwards Compatibility
+=======================
+
+Apart from the :class:`~typing.TypedDict` internal implementation change, this
+PEP does not bring any backwards incompatible changes.
+
+
+Security Implications
+=====================
+
+There are no known security consequences arising from this PEP.
+
+
+How to Teach This
+=================
+
+The new inlined syntax will be documented both in the :mod:`typing` module
+documentation and the :ref:`typing specification <typing:typed-dictionaries>`.
+
+As mentioned in the `Rationale`_, it should be mentioned that inlined typed
+dictionaries should be used for small structured data to not hurt readability.
+
+
+Reference Implementation
+========================
+
+Mypy supports a similar syntax as an :option:`experimental feature <mypy:mypy.--enable-incomplete-feature>`::
+
+    def test_values() -> {"int": int, "str": str}:
+        return {"int": 42, "str": "test"}
+
+Pyright has added support in version `1.1.297`_ (using :class:`dict`), but was later
+removed in version `1.1.366`_.
+
+.. _1.1.297: https://github.com/microsoft/pyright/releases/tag/1.1.297
+.. _1.1.366: https://github.com/microsoft/pyright/releases/tag/1.1.366
+
+Runtime implementation
+----------------------
+
+A draft implementation is available `here <https://github.com/Viicos/cpython/commit/49e5a83f>`_.
+
+
+Rejected Ideas
+==============
+
+Using the functional syntax in annotations
+------------------------------------------
+
+The alternative functional syntax could be used as an annotation directly::
+
+    def get_movie() -> TypedDict('Movie', {'title': str}): ...
+
+However, call expressions are currently unsupported in such a context for
+various reasons (expensive to process, evaluating them is not standardized).
+
+This would also require a name which is sometimes not relevant.
+
+Using ``dict`` with a single type argument
+------------------------------------------
+
+We could reuse :class:`dict` with a single type argument to express the same
+concept::
+
+    def get_movie() -> dict[{'title': str}]: ...
+
+While this would avoid having to import :class:`~typing.TypedDict` from
+:mod:`typing`, this solution has several downsides:
+
+* For type checkers, :class:`dict` is a regular class with two type variables.
+  Allowing :class:`dict` to be parametrized with a single type argument would
+  require special casing from type checkers, as there is no way to express
+  parametrization overloads. On ther other hand, :class:`~typing.TypedDict` is
+  already a :term:`special form <typing:special form>`.
+
+* If fufure work extends what inlined typed dictionaries can do, we don't have
```

Comment:
Did you consider reusing `typing.Dict` here? I can see some potential advantages. It's less verbose, doesn't have the baggage of `builtins.dict`, and the fact that this is a "typed" dict is already implicit given that it is imported from `typing` and involves type annotations.

Reply:
Showing my TypeScript bias here :sweat_smile: but I'm wondering how sacrilegious it would be to  avoid the subscription all together and simply use the dictionary itself?

```python
def get_movie() -> {'name': str, 'year': int}:
    return {
        'name': 'Blade Runner',
        'year': 1982,
    }
```

Reply:
This causes problems with the `|` operator at least. I've been thinking of writing up something like "Why can't we ...?" discussing why certain "obvious" ways to make typing more ergonomic may not work well in practice.

Reply:
I think that such writeup would be very useful!

Reply:
I added an extra section in the rejected ideas regarding the plain dict syntax.

On using `typing.Dict`, I kind of like the idea for the reasons you mentioned @erictraut. `Dict[{...}]` implicitly spells out as something like:

```
Dict[{...}]
|      |------> ...with some typed fields
|--> a dict... 
```

However, I still think this is debatable for a couple reasons:
- [`typing.Dict`](https://docs.python.org/3/library/typing.html#typing.Dict) is currently marked as deprecated in the Python documentation (although not scheduled for removal). It might be confusing to undeprecate it.
- The same subscripting syntax — with the only difference being the number of type arguments — means two different things. Can be confusing as well.

I'll add an entry to the open issues so that we can further discuss this option.